### PR TITLE
Store EPICS connection status in file

### PR DESCRIPTION
**Pull request by mchorazak**
### Issue

DM-411

### Description of work
Related to [this](https://github.com/ess-dmsc/forward-epics-to-kafka/pull/289) forwarder PR.

- Created new `ep00` writer module that handles messages. 
- Connection status and it's timestamp is written `connection_status` and `connection_status_time` datasets respectively

Covered by unit and system tests.

### Nominate for Group Code Review

- [ ] Nominate for code review

### CODE REVIEW

**Comment by matthew-d-jones** [comment]
Once https://github.com/ess-dmsc/forward-epics-to-kafka/pull/292 is merged these data will be forwarded as part of `f142` payloads rather than in separate `ep00` messages. I will close this PR and cherry pick anything useful from it into a new branch.

**Comment by matthew-d-jones** [comment]
Reinstating PR, I got confused due to the dataset naming. Connection status data are independent of EPICS alarms. Dataset names used here should be `connection_status` and `connection_status_time`.

File: src/schemas/ep00/FlatbufferReader.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#pragma once
+
+#include "../../FlatbufferReader.h"
+#include <flatbuffers/flatbuffers.h>
```
Comment (mattclarke) [inline src/schemas/ep00/FlatbufferReader.h]:
```suggestion
#include "flatbuffers/flatbuffers.h"
```
I think that seems more correct, but happy to be overruled

File: src/schemas/ep00/FlatbufferReader.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace FileWriter {
+namespace Schemas {
+namespace ep00 {
+#include <ep00_epics_connection_info_generated.h>
```
Comment (mattclarke) [inline src/schemas/ep00/FlatbufferReader.h]:
```suggestion
#include "ep00_epics_connection_info_generated.h"
```

File: src/schemas/ep00/ep00_rw.cpp [hunk 3]
```diff
@@ -0,0 +1,63 @@
+#include "ep00_rw.h"
+#include <HDFFile.h>
```
Comment (mattclarke) [inline src/schemas/ep00/ep00_rw.cpp]:
```suggestion
#include "HDFFile.h"
```

File: src/schemas/ep00/ep00_rw.cpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+#include <json.h>
```
Comment (mattclarke) [inline src/schemas/ep00/ep00_rw.cpp]:
```suggestion
#include "json.h"
```

File: src/schemas/ep00/ep00_rw.h [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+public:
+  InitResult init_hdf(hdf5::node::Group &HDFGroup,
+                      std::string const &HDFAttributes) override;
+  void parse_config(std::string const &ConfigurationStream) override;
+  HDFWriterModule::InitResult reopen(hdf5::node::Group &HDFGroup) override;
+  void write(FlatbufferMessage const &Message) override;
+
+  int32_t close() override;
+
+  HDFWriterModule();
```
Comment (mattclarke) [inline src/schemas/ep00/ep00_rw.h]: Can be defaulted

File: src/schemas/ep00/ep00_rw.cpp [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  auto FlatBuffer = get_fbuf(Message.data());
+  std::string Status = EnumNameEventType(FlatBuffer->type());
+  StatusDataset.appendString(Status);
+  auto FBTimestamp = FlatBuffer->timestamp();
+  TimestampDataset.appendElement(FBTimestamp);
+}
+
+int32_t HDFWriterModule::close() { return 0; }
+
+HDFWriterModule::HDFWriterModule() {}
```
Comment (mattclarke) [inline src/schemas/ep00/ep00_rw.cpp]: Default

File: src/tests/EpicsConnectionStatusTests.cpp [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+}
+
+class Schema_ep00 : public ::testing::Test {
+public:
+  void SetUp() override {
+    try {
+      FileWriter::FlatbufferReaderRegistry::Registrar<
+          FileWriter::Schemas::ep00::FlatbufferReader>
+          RegisterIt("ep00");
+    } catch (...) {
```
Comment (mattclarke) [inline src/tests/EpicsConnectionStatusTests.cpp]:
```suggestion
    } catch (std::runtime_error const &error) {
```

File: src/tests/EpicsConnectionStatusTests.cpp [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+  auto Datatype = hdf5::datatype::String::fixed(20);
+  Datatype.encoding(hdf5::datatype::CharacterEncoding::UTF8);
+  auto Dataspace = hdf5::dataspace::Simple({1});
+  EXPECT_NO_THROW(StatusDataset.read(StatusData, Datatype, Dataspace));
+  std::string StringFromDataset = StatusData[0];
+  removeTrailingNullFromString(StringFromDataset);
+  EXPECT_EQ(StringFromDataset, "CONNECTED");
+}
+
+TEST_F(Schema_ep00, FBReaderNoSourceName) {
```
Comment (mattclarke) [inline src/tests/EpicsConnectionStatusTests.cpp]: Test name doesn't explain what is being tested.
It is actually quite difficult to work out what the test is doing from the code.
Looks a bit like more than one test.

File: src/tests/EpicsConnectionStatusTests.cpp [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+  EXPECT_TRUE(TestGroup.has_dataset(TimestampName));
+}
+
+TEST_F(Schema_ep00, ReopenFile) {
+  ep00::HDFWriterModule Writer;
+  EXPECT_FALSE(Writer.reopen(UsedGroup) ==
+               HDFWriterModule_detail::InitResult::OK);
+}
+
+TEST_F(Schema_ep00, FileInitFail) {
```
Comment (mattclarke) [inline src/tests/EpicsConnectionStatusTests.cpp]:
```suggestion
TEST_F(Schema_ep00, FileInitFailsIfInitialisedTwice) {
```

File: src/tests/EpicsConnectionStatusTests.cpp [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+  std::unique_ptr<std::int8_t[]> Buffer =
+      GenerateFlatbufferData(BufferSize, Timestamp, Status, SourceName);
+  ep00::HDFWriterModule Writer;
+  EXPECT_TRUE(Writer.init_hdf(UsedGroup, "{}") ==
+              HDFWriterModule_detail::InitResult::OK);
+  EXPECT_TRUE(Writer.reopen(UsedGroup) ==
+              HDFWriterModule_detail::InitResult::OK);
+  FileWriter::FlatbufferMessage TestMsg(
+      reinterpret_cast<const char *>(Buffer.get()), BufferSize);
+  EXPECT_NO_THROW(Writer.write(TestMsg));
```
Comment (mattclarke) [inline src/tests/EpicsConnectionStatusTests.cpp]: I would only use EXPECT_NO_THROW if that was the point of the test/API, e.g. testing that something doesn't throw under certain conditions.
This seems like testing if just for the sake of it

Comment (mattclarke) [reply src/tests/EpicsConnectionStatusTests.cpp]: Also applies to the other tests

File: src/tests/EpicsConnectionStatusTests.cpp [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+
+TEST_F(Schema_ep00, FBReaderNoSourceName) {
+  size_t BufferSize;
+  uint64_t Timestamp = 5555555;
+  std::string SourceName;
+  auto Status = EventType::NEVER_CONNECTED;
+  std::unique_ptr<std::int8_t[]> Buffer =
+      GenerateFlatbufferData(BufferSize, Timestamp, Status, SourceName);
+  ep00::HDFWriterModule Writer;
+  EXPECT_TRUE(Writer.init_hdf(UsedGroup, "{}") ==
```
Comment (mattclarke) [inline src/tests/EpicsConnectionStatusTests.cpp]: This is repeating the expectations from FileInitOk and ReopenFileSuccess.
That is not usually a good thing.

File: src/tests/EpicsConnectionStatusTests.cpp [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+TEST_F(Schema_ep00, WriteDataSuccess) {
+
+  size_t BufferSize;
+  uint64_t Timestamp = 5555555;
+  std::string SourceName = "SIMPLE:DOUBLE";
+  auto Status = EventType::CONNECTED;
+  std::unique_ptr<std::int8_t[]> Buffer =
+      GenerateFlatbufferData(BufferSize, Timestamp, Status, SourceName);
+  ep00::HDFWriterModule Writer;
+  {
```
Comment (mattclarke) [inline src/tests/EpicsConnectionStatusTests.cpp]: This is repeating the expectations from FileInitOk and ReopenFileSuccess.
That is not usually a good thing.

File: src/schemas/ep00/FlatbufferReader.h (hunk 1 above)
Comment (matthew-d-jones) [reply src/schemas/ep00/FlatbufferReader.h]: `flatbuffers.h` is an external library header, so I think `<>` is fine in this case, but I agree with you on some of the other cases you've pointed out.

File: src/tests/EpicsConnectionStatusTests.cpp (hunk 8 above)
Comment (matthew-d-jones) [reply src/tests/EpicsConnectionStatusTests.cpp]: It seems to be documenting that status data are still written if the source name is missing from the message, but I think I'll remove the test as messages without a source name never reach the writer module anyway.

**Pull request merged**